Workbox Webpack Plugin v3

goldhand opened the pull request:
## tl;dr

This is a prototype for the `workbox-webpack-plugin` for the v3 release. This address some of the points mentioned in #696 . Not all of the configuration from the existing `workbox-webpack-plugin` has been migrated (or configuration from `sw-precache-webpack-plugin`) but that was not the focus of this prototype. The goal of this prototype is to solve for some of the difficult webpack-specific features, including:
* Supporting `webpack-dev-middleware` by generating the `service-worker`, `file-manifest`, and built version of `workbox-sw` **in-memory** (rather than writing to disk). I've tested this and confirmed support for `webpack-dev-middleware` / `webpack-dev-server`.
* Filtering files by chunk names.
* Copying`workbox-sw` to the webpack assets.


⚠️ *I haven't updated the tests yet, I was manually testing by using an ejected create-react-app instance that imports the plugin using a relative file path.*
⚠️  *One issue (mentioned in the [blocks](#blocks) section below, was that I wasn't able to import the `workbox-sw` package into the `workbox-webpack-plugin` package. See details in the blocks section for my work-around.*


# V3 Changes to workbox-webpack-plugin

Following are details of the design, proposed api changes, and issues I am having that I am hoping someone will provide help with.

## Requirements
*from the #725 v3 tracker*
* [ ] Switch to CDN by default
* [ ] Include option to not use CDN and copy files over
* [ ] Update API to implement v3 proposal
* [x] Use compilation assets to generate a manifest (instead of globPatterns)
* [x] Allow user to specify chunks to be used when creating the manifest
* [x] Generate service worker “in-memory” and make webpack responsible for writing to file system
* [ ] Support sw-precache config in workbox-webpack plugin (runtime-caching? drop staticFileIgnores in favor of “chunks” api)

Im not clear what the *CDN* related tasks are.

## Re-architecture
The objective is to provide a seamless experience for developers who want to use workbox in projects that rely on webpack for building.
Webpack contains information about the build files and versions of those files that workbox can leverage to manage service worker caching.

### Tenets
* Try to use as many existing workbox utilities to generate the service worker through webpack without compromising the developer experience.
* Ensure all required service worker assets are emitted through webpack so workbox can be used with `webpack-dev-middleware`.

### Design
* In most webpack build systems, webpack is aware of all the built static assets as well as if there are changes in those assets. Were going to use webpack to generate a file manifest that `workbox-sw` can use for precaching instead of using `workbox-build/generate-file-manifest`.
* Filter manifest files to include / exclude files by chunk name. This is provided by the `config.chunks` and `config.excludeChunks`, respectively (mirroring the `html-webpack-plugin` api).
* If we need to use the `compiler.inputFileSystem._readFile` method instead of `fs.readFile`, use the `workbox-webpack-plugin/utils/read-file.js` method that is configured to use `compiler.inputFileSystem._readFile` during the **run** phase of the webpack compilation lifecycle.
* `workbox-build` will copy a built `workbox-sw` into the service worker output path. This needs to done through the webpack. During the **make** phase of the webpack compilation, we use `workbox-webpack-plugin/utils/copy-workbox-sw` to add a built version of `workbox-sw` to the webpack compilation assets array. We can't use the `workbox-build/utils/copy-workbox-sw` utility because it writes to the filesystem.
* The version of `workbox-sw` is included in it's filename so we need that information to import it in the generated service worker. `workbox-webpack-plugin/utils/copy-workbox-sw` also returns the `workbox-sw` filename so we can pass it to `workbox-webpack-plugin/generate-webpack-sw`.
* During the **emit** phase of the webpack lifecycle, we:
  1. Get the manifest entries
  1. Use the entries to generate a file-manifest
  1. Generate a service worker with the file-manifest name and `workbox-sw` filename or copy a service worker from the `config.swSrc` then prepend it with the required `importScripts(workbox-sw.js, file-manifest.js)`.
  1. Add both the file-manifest and the service-worker to the webpack assets.

#### Todo
* I'm not passing any configuration into `sw-build.generateSW` as we would need to. This is because of the pattern where utilities *write* to the file system. If we don't change the utilities, one way to fix this would be to copy the *template populate* logic from `sw-build/generate-sw`. Ideally we could abstract the populating logic from `generate-sw` into a util that does not write to the file system.
* Honor `swDest` incase users want to separate the service worker from build directory.
* There may be cases where assets are not emitted by webpack (the `public/` dir in create-react-app, for example). We should provide a way to "merge" `workbox-build/get-file-manifest-entries` with `workbox-webpack-plugin/get-webpack-manifest-entries`.  `workbox-build/get-file-manifest-entries` supports globPatterns for finding assets, and can be used to generate manifest entries for files that do not go through webpack. Could be a simple `Object.assign(webpackEntries, workboxEntries)`.
* Supoort `sw-precache-webpack-plugin` configurations.
* Fix tests (once the `sw-build/generate-sw`) item is resolved.

### API changes
These are just proposals, we can stay within the exisiting API but I wanted to collect feedback for these changes.

* Running `workbox-build` in `injectManifest` mode will not copy `workbox-sw` over but this is a known pain point for developers. Instead, copy the `workbox-sw` over then prefix the existing service worker with an `importScripts(workbox-sw-V.xxx.js)` line.
* Always generate a file manifest (do not support `injectManifest` mode). We can effectively "inject" the manifest by taking a `config.manifestVarName` parameter that will be used in the `file-manifest.js` template: (`self.${config.manifestVarName} = ${fileManifest}`) and prepend an `importScripts(${manifestFileName})` line to the existing service worker.

### Blocks
* I wasn't able to import the `workbox-sw` package into the `workbox-webpack-plugin` package (in the `workbox-webpack-plugin/utils/copy-workbox-sw` utility). I kept getting a "module not found error" when I manually tested using the create-react-app approach). My work around was to add the `copy-workbox-sw` to the `workbox-build` package (that I was able to import) and manually test using that.
* A lot of the utilities use `fs` to read lodash templates. There may be problems using `fs` in webpack plugins (I removed it from `sw-precache-webpack-plugin` and @TheLarkInn mentioned using `compiler.outputFileSystem` instead).
* Utilities often *populate* then *write* the populated template to the filesystem. I'm having to duplicate and isolate the *populate* logic and am referencing the lodash templates across modules.

goldhand commented:
I signed it!

jeffposnick commented:
Thanks so much, @goldhand!

I'll give it a thorough review in a bit.

Some housekeeping: we'll want to target the `v3` branch with this PR, as we still might cut a few releases using the v2 major version number from `master` before we're ready to launch any breaking changes. At the top of my TODO list is to port over the `workbox-build` module to the `v3` branch, which should then make it easier for you to land your code in `v3` as well.

addyosmani commented:
>  At the top of my TODO list is to port over the workbox-build module to the v3 branch, which should then make it easier for you to land your code in v3 as well.

Was just talking to @gauntface about this. Exciting to see this slowly coming together, @goldhand. Thanks for getting this work started!

goldhand commented:
Cool, I'll move this to target the v3 branch. This is still very "WIP", just wanted to get something out there to see if anyone could help in areas.

gauntface commented:
Thanks for this work @goldhand.

I'm very keen to get all the changes to workbox-build known for V3 so this is very timely for @jeffposnick's work.

## Blocks

- workbox-sw would need to be added as a dependency to the webpack plugin to be able to copy the files over. This will likely change in V3 (but there will still be copying done).
    - Basically workbox-sw currently contains all the code from a number of modules. In V3 these will be seperate files and developers only need to pull in what they use.
- Why are there problems reading files in a webpack plugin (sorry if that's a stupid question)? When generating a service worker, we'll need to grab an example template and alter it. `compiler.outputFileSystem` sounds weird to me in that the reading is done before anything is outputted. Would it make send to alter workbox-build to use some other file reader API or would we need to switch to a template which is a string and manipulate that rather than read a file?
- What would an ideal API surface in workbox-build be to prevent you from needing to duplicate the logic? Would it make sense to add API's to workbox-build that would return a string (in the case of generating a SW) and an API that accepted and returns strings as taking in a service worker file and returning an altered service worker file.

# API Changes

I think it would be good to have a brainstorm on fixing this. There are a number of issues around this space and a number of possible solutions, all with varying degrees of magic / flexibility and usefulness. 

@jeffposnick Do you think you could work on a proposal for workbox-build that would address some of the concerns raised above? If not I'm more than happy to have a crack at it, but I know you are owning that module and I think having a fresh pair of eyes on it would be good. Also happy if you'd like to have a discussion of possible ideas and approaches before attempting a proposal. Basically don't want to block and don't want to waste your time.

I think a proposal might be helpful before starting work.

#### What is the CDN?

@goldhand In V3 we are planning on supporting a CDN. This should make it easy for developers to use Workbox on Glitch and also avoid the issue of needing to know where workbox-sw is in `node_modules` + it avoids having magic to inject stuff in a service worker script (as the cdn import can be copy and pasted).

gauntface commented on packages/workbox-webpack-plugin/build.js:
```diff
@@ -17,6 +17,6 @@ const fse = require('fs-extra');
 module.exports = () => {
-  return fse.copy(path.join(__dirname, 'index.js'),
-    path.join(__dirname, 'build', 'index.js'));
+  return fse.copy(path.join(__dirname, 'src'),
```
Is this needed?

I think this may have been a thing that was needed for older versions but the build model has changed in V3. At the moment lerna will just publish the directoy as is and won't use build at all.

jeffposnick replied:
I've got a build process working in v3 for Node modules is this PR: https://github.com/GoogleChrome/workbox/pull/811

(As of right now, it determines that something is a Node module based on the presence of `engines.node` in `package.json`, which is a bit of magic.)

In summary, assuming #811 is merged first and this PR is rebased against the `v3` branch, you should be able to get rid of this `build.js` file entirely.

jeffposnick replied:
The goal is to get #811 reviewed and merged into `v3`, yes. So that would ultimately be what this PR would build on top of. I'll modify the code in that PR to address some of the things you've asked for.

addyosmani replied:
With #811 now merged into v3, now might be a good time to try rebuilding this PR on top of what is there. It's unclear to me how much churn we might see in the workbox-build interface changes but @jeffposnick seemed to be supportive of either rebasing against v3 or waiting a little longer. I think we're fine with whatever your preference is, @goldhand.

gauntface commented on packages/workbox-webpack-plugin/package.json:
```diff
@@ -23,7 +23,8 @@
   "dependencies": {
-    "workbox-build": "^2.0.1"
+    "workbox-build": "^2.0.1",
+    "workbox-sw": "^2.0.1"
```
It's worrying if you added dep for workbox-sw and it can't be copied over :-S

jeffposnick replied:
I wonder whether the issue you described was due to `lerna`-being-`lerna`.

If you do a `gulp lerna-bootstrap` in the top-level `workbox/` directory that might take care of ensuring that `workbox-build`'s dependency on `workbox-sw` gets satisfied, without you having to explicitly list it.

gauntface commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,108 @@
+   * @private
+   * @param {Object} [compiler] default compiler object passed from webpack
+   */
+  apply(compiler) {
```
Could you add some comments about the lifecycle of the compiler callbacks?

Reading through this code I'm guessing a lot about where things occur.

gauntface commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,108 @@
+        // Add the workbox-sw file to compilation assets
+        compilation.assets[workboxSWName] = webpackAsset(workboxSW);
+        compilation.assets[`${workboxSWName}.map`] = webpackAsset(workboxSWMap);
+        this.config.set('workboxSWFilename', workboxSWName);
```
Could this be this._config?

gauntface commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,108 @@
+   * @private
+   * @return {Object} The configuration for a given compilation.
+   */
+  get config() {
```
Why are you following this model of exporting an object that seems to read and write to the config option?

addyosmani replied:
I had a similar question. I thought that this might be trying to replicate a pattern used in the sw-precache webpack plugin, but I don't think this model in this form was used there (I could be wrong) :)

goldhand replied:
No need for this! Will completely change this "config" when I clean up the api

addyosmani replied:
👍

gauntface commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,108 @@
+        manifestVarName = '__file_manifest',
+        workboxSWFilename,
+        swSrc,
+      } = config;
```
I've never seen this syntax before, what is this doing?

addyosmani replied:
@gauntface This is a form of loose destructuring assignment where properties can be unpacked from an object and assigned to a variable with a different name than the object property.

goldhand replied:
This is destructuring with "defaults".

So `var manifestVarName = config.manifestVarName || '__file_manifest'`;

gauntface commented on packages/workbox-webpack-plugin/src/lib/generate-webpack-manifest.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = (manifestEntries, manifestVarName) =>
```
Nit: This file should probably be named something else. `generate-webpack-manifest` makes it sound like a webpack thing rather than a file manifest that is used in the service worker.

addyosmani replied:
> generate-manifest-with-webpack

This works!

gauntface commented on packages/workbox-webpack-plugin/src/lib/generate-webpack-sw.js:
```diff
@@ -0,0 +1,29 @@
+const {readFile} = require('./utils/read-file');
+
+const importScripts = ({
```
Not: I'd rename this function. importScripts is way to generic / confusing given the context. `wrapAsImport` or something in that direction would be helpful.

gauntface commented on packages/workbox-webpack-plugin/src/lib/generate-webpack-sw.js:
```diff
@@ -0,0 +1,29 @@
+  manifestFilename,
+}) => `importScripts('${workboxSWFilename}', '${manifestFilename}');`;
+
+const generateSW = ({
```
Again, given that workbox-build has a generateSW method and workbox-cli has generate-sw, the naming is confusing and I'm not sure what this function is doing or why it's doing it.

goldhand replied:
@gauntface the problem with using `workbox-build`'s `generateSW` method is that it writes to the file system. The reason I have any of these `generate-template` utils is because I couldn't figure out how to use the existing code without writing to the filesystem.

This `generate-webpack-sw` util is a placeholder intended to highlight one of the issues described in the **blocks** section:

> Utilities often populate then write the populated template to the filesystem. I'm having to duplicate and isolate the populate logic and am referencing the lodash templates across modules.

Maybe I missed it, but does `workbox-cli` have a way to generate the template without writing to fs?

*sidenote: sorry for the single comments. Will answer the other points in a review later*

jeffposnick replied:
See https://github.com/GoogleChrome/workbox/pull/811/files#r140564059 for what we can do in v3 of`workbox-build`, if that would help.

gauntface commented on packages/workbox-webpack-plugin/src/lib/generate-webpack-sw.js:
```diff
@@ -0,0 +1,29 @@
+const {readFile} = require('./utils/read-file');
```
Is this the webpack templating that you've used instead of workbox-builds templating?

If so, renaming this to `webpack-generate-sw-template` would help with context.

addyosmani replied:
In favor of the naming Matt suggested above.

gauntface commented on packages/workbox-webpack-plugin/src/lib/generate-webpack-sw.js:
```diff
@@ -0,0 +1,29 @@
+
+const generateOrCopySW = (config, swSrc) => new Promise((resolve, reject) => {
+  if (!swSrc) {
+    return generateSW(config).then((serviceWorker) => resolve(serviceWorker));
```
Nit: Is `serviceWorker` a generated string or a file path?

gauntface commented on packages/workbox-webpack-plugin/src/lib/generate-webpack-sw.js:
```diff
@@ -0,0 +1,29 @@
+    return generateSW(config).then((serviceWorker) => resolve(serviceWorker));
+  } else {
+    return readFile(swSrc).then((serviceWorkerSource) =>
+      resolve(`${importScripts(config)}\n${serviceWorkerSource}`)
```
I think the naming would help with this.

gauntface commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+  );
+
+/**
+ * Maps webpack asset filenames to their chunk hash
```
Could you add a brief description of the input and outputs of this function. Would help read the code.

goldhand replied:
yes

gauntface commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+  if (includeNames || excludeNames) {
+    // Only include chunks in config.chunks and
+    // exclude any chunks in config.excludeChunks
+    chunks = filterChunks(chunks, includeNames, excludeNames);
```
Would it be accurate to do the following:

Rename `chunks` to `allChunks` or `allCompilationChunks`, renaming `includeNames` to `chunksToInclude` and `excludeNames` to `chunksToExclude` and then add a variables `chunksToPrecache` would would result in:

```javascript
let chunksToPrecache = allChunks;
if (chunksToInclude || chunksToExclude) {
    chunksToPrecache = filterChunks(allChunks, chunksToInclude, chunksToExclude);
}
```

gauntface commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+   *
+   * @type {Array<string>}
+   */
+  const manifestFiles = includeNames
```
I'm kinda confused with all the alterations and this object keys here, would this change to:

```javascript
const assetsToChunkHash = mapAssetsToChunkHash(precacheChunks);
const manifestFiles = Object.keys(precacheChunks);
```

This being due to the check includeNames || excludeNames above rendering this if statement no longer needed.

goldhand replied:
yeah it's kind of confusing. I'll update comments to clarify more.

`assetsToChunkHash` is a map of `filename`: `chunkhash` and only contains assets that belong to a chunk with a name property. It's also filtered out chunks according to `include` / `exclude`.

`compilation.assets` is all assets that belong to the webpack compilation, even assets that belong to chunks with no name.

gauntface commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+    ), {});
+
+const generateManifestEntries = (compiler, compilation, config) => {
+  const {hash} = compilation; // compilation hash
```
You could inline the single use of hash to the bottom of this function as `compilation.hash`.

goldhand replied:
sure!

addyosmani replied:
Thanks!

gauntface commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+const url = require('url');
```
To prevent developers from precaching huge files (i.e. videos / audio files) we have a size limit on the files we'll precache. Is that possible with WebPack?

addyosmani replied:
Perhaps something like https://github.com/ai/size-limit/blob/master/index.js#L135 would do the trick?

addyosmani replied:
TBH I think it was static and something we internally enforced. @jeffposnick do you know if this was something we otherwise made configurable?

jeffposnick replied:
It's configurable via the `maximumFileSizeToCacheInBytes` option.

gauntface commented on packages/workbox-webpack-plugin/src/lib/utils/read-file.js:
```diff
@@ -0,0 +1,23 @@
+ * the run phase of the webpack compilation lifecycle by passing the function
+ * to the `setReadFile` function.
+ */
+let readFileFn;
```
I'm not crazy about this pattern but I can see why you've done it.

goldhand replied:
Yeah, me either. This is just incase we can't use `fs` or `fse` in the webpack compilation. Is there a better pattern for this you would suggest?

gauntface commented on packages/workbox-webpack-plugin/src/lib/utils/webpack-asset.js:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Creates a webpack asset from a string that can be added to a compilation
```
Not sure what this does or why it's needed, could you add some more context? Sorry I'm such a webpack noob :(

addyosmani replied:
I understand the reason this is needed, but agree for future contributors it would be useful to expand on why this is needed.

goldhand replied:
I will elaborate what this is in the docs.

These are the minimum properties that webpack expects a single asset to have. Given a string (`asset`) we can structure it for webpack without any other arguments.

addyosmani commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+ * @param {Array<Object>} chunks Webpack chunks
+ * @param {Array<string>} include Chunk names to include
+ * @param {Array<string>} exclude Chunk names to exclude
+ * @return {Array<Object>} Filtered array of chunks
```
Once we have some docs together (for a more fleshed out version), it would be great for them to capture a few different examples of using chunks, include and exclude.

goldhand replied:
+1

jeffposnick commented on packages/workbox-webpack-plugin/package.json:
```diff
@@ -23,7 +23,8 @@
+    "workbox-build": "^2.0.1",
+    "workbox-sw": "^2.0.1"
   },
   "author": "Google's Web DevRel Team",
```
Feel free to add your name!

addyosmani replied:
Yeah. We definitely want you to feel like as much a part of the team as anyone else. Feel free to add yourself :)

@jeffposnick @gauntface at some point, we might want to change this to "Workbox Core Team" or reference an owners list we can also add folks like Will to.

jeffposnick commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * Creates an instance of WorkboxBuildWebpackPlugin.
+   *
+   * @param {Object} [config] All the options as passed to `workbox-build`.
```
`{@link module:workbox-build~Configuration}` should, if I understand my JSDoc syntax, generate a link to the canonical description of all the `workbox-build` options.

jeffposnick commented on packages/workbox-webpack-plugin/src/lib/generate-webpack-manifest.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = (manifestEntries, manifestVarName) =>
+  new Promise((resolve) => resolve(
```
Nit: You can use `Promise.resolve(value)` as shorthand to create a promise that resolves with a given value. No need to explicitly construct a `Promise` and pass it a function.

jeffposnick commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+        ...chunk.files.map((f) => ({[f]: chunk.renderedHash}))
+    ), {});
+
+const generateManifestEntries = (compiler, compilation, config) => {
```
Something I noticed when running your code locally was that the `workbox-sw` (and `.map`) file were getting picked up here:

```
  {
    "revision": "9a0110a4f8c1796a1414",
    "url": "/workbox-sw.prod.v2.0.1.js"
  },
  {
    "revision": "9a0110a4f8c1796a1414",
    "url": "/workbox-sw.prod.v2.0.1.js.map"
  },
```

While including these files in the manifest won't break anything (the browser will ignore them when performing `importScripts()`), it means that there's some extra bytes transferred and stored on disk during SW installation, and it would be preferable to exclude them.

goldhand replied:
Oh, I thought we would want to cache these? I am curious, how come we don't want to cache these assets?

jeffposnick replied:
Anything pulled in with `importScripts` (e.g. `workbox-sw.prod.v2.0.1.js`) is implicitly cached and available offline—this behavior is defined in the service worker specification ([section 7.3.2](https://w3c.github.io/ServiceWorker/#importscripts), item 3.1). Requests for those URLs don't trigger a service worker `fetch` event.

The `.map` isn't covered by this policy, but I don't actually know that the request for the `.map` made by DevTools while debugging actually triggers a `fetch` event either, since DevTools might not be considered an active client of the service worker. And even if the `.map` request does go through the service worker, needing to use the sourcemap while offline is an edge case that wouldn't warrant the extra bytes transferred and stored.

addyosmani replied:
> needing to use the sourcemap while offline is an edge case that wouldn't warrant the extra bytes transferred and stored.

I strongly agree with this. I've seen plenty of examples of folks also generating and accidentally caching `.map` files that were pretty large (1MB) causing a perf hit when it's easier to just not cache this asset.

jeffposnick commented on packages/workbox-webpack-plugin/src/lib/utils/copy-workbox-sw.js:
```diff
@@ -0,0 +1,19 @@
+const {readFile} = require('./read-file');
+const errors = require('workbox-build/src/lib/errors');
+
+module.exports = (readFileFn) => {
```
FWIW, in the v3 branch, we're transpiling all the Node code, so feel free to use features like `async`/`await` if you'd prefer.

goldhand replied:
YAY!

jeffposnick commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+
+const entry = (revision, filePath, publicPath) => ({
+  revision,
+  url: url.resolve(publicPath, filePath),
```
I remember this coming up in a review of the earlier revision of the Webpack plugin: `publicPath` [is defined](https://webpack.js.org/configuration/output/#output-publicpath) as something that's prepended to every URL, not joined with it, so this should probably be `publicPath + filePath`;

goldhand replied:
Oops, me too. Will change that

jeffposnick commented on packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js:
```diff
@@ -0,0 +1,76 @@
+const url = require('url');
+
+const entry = (revision, filePath, publicPath) => ({
+  revision,
```
One thing that bears mentioning: as an optimization, the `precache()` function will not append a cache-busting query parameter when populating its caches if the `revision` field is not present.

That's the same behavior that in `sw-precache-webpack-plugin` was defined by the `dontCacheBustUrlsMatching` configuration option, for which we recommended using a RegExp that matches URLs that contained hashes in them already.

I'm not sure exactly how, but is there any indication at this point in the build process whether `filePath` contains a versioning hash in it? (I'd assume it wouldn't for HTML files, but would for most other assets.) If you can safely infer the presence of a hash, then you can leave out `revision` and get optimal behavior without developers needing to use `dontCacheBustUrlsMatching`.

goldhand replied:
Makes sense, we can check if they are using a version hash in the chunk name and filter those out. Then just ignore the `dontCacheBustUrlsMatching` option altogether?

jeffposnick replied:
If someone goes out of their way to supply `dontCacheBustUrlsMatching` then perhaps it should still be honored, but being able to implicitly provide that optimization is definitely more developer-friendly.

I think it's more of a case of you not taking specific action one way or another in your Webpack-specific code.

jeffposnick commented:
@goldhand, as an FYI, #811 is now merged, so `workbox-build` is in the `v3` branch. The codebase at the moment is functionally equivalent to what was in `master`, so it doesn't have the modified interfaces we've been discussing (tracked in #803), but it does bring into `v3` testing setup and build steps for `node` projects, which were previously missing.

It should be possible to rebase against `v3` at this point, in advance of waiting for the `workbox-build` interfaces to be updated. Or you could wait for those interface changes, and do it all in one go. Let us know if you run into any rough patches.

addyosmani commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,108 @@
+ *    new WorkboxBuildWebpackPlugin({
+ *      chunks: ['main'],
+ *      excludeChunks: ['mutable'],
+ *      manifestFilename: string,
```
Thanks. Even including the description you provided above in comments would be more than sufficient to cover future contributors.

jeffposnick commented:
https://github.com/GoogleChrome/workbox/issues/803#issuecomment-333951970 has an update on the current state of `workbox-build` in `v3`, with an eye on features that should simplify this PR.

goldhand commented:
Thanks @jeffposnick I will be working on this over the weekend! generate-sw.js is exactly what I needed and sweet `async/await`

anilanar commented:
I'm late to the party, I'd love to review this PR and test it against our complicated webpack configuration @trivago.

goldhand commented:
Thanks @anilanar ! Would appreciate that. Should have a more complete working version after this weekend. Here is the webpack-related work that will support running with webpack-dev-server. Need to bring in more of the workbox functionality now so if you want to add to this pr or use as a starting point for another pr, please do!

goldhand commented:
@jeffposnick I made a decent amount of progress this weekend:

* Migrated to `v3` branch
* Used `workbox-build` utilities where I could
* Updated documentation where requested in feedback
* Renamed files / functions so it was clearer what they do
* Noted missing todo's in code comments with links to the review comments for details (in case someone else wants to make these fixes).
* Added a method that returns all configuration params for `generateSWString`. I am highlighting this because the idea was we could explicitly specify whatever properties WorkboxWebpack plugin config should pass to `generateSWString`. This allow should allow us to easily customize the api / experience.

Going to call it a night but I'll try to update with more details / respond to any questions on this pr over this next week.

goldhand commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,181 @@
+   * @return {Object} All workbox configuration options that can be accepted
+   * by {@link module:workbox-build.generateSWString}
+   */
+  get generateSWStringOptions() {
```
@jeffposnick this is where we should specify which generateSWString properties should be supported.

anilanar commented on packages/workbox-webpack-plugin/src/lib/utils/resolve-webpack-url.js:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @memberof module:workbox-webpack-plugin
+ */
+module.exports = (...paths) => paths.join('');
```
This would get you in trouble, because all webpack plugins I've used so far are capable of handling public paths that do not end with a `/`, including webpack itself.

goldhand replied:
There was discussion about this. This is why I isolated the path logic too. Do you think we should do a `url.resolve` rather that `String.join`?

anilanar replied:
Maybe keep it as it is. It is quite obvious what's wrong when you see a slash missing between publicPath and filename.

jeffposnick replied:
Yes, I'm the one who keeps making an issue out of this 😄 

It seems like we should be following the conventions from the Webpack docs, which says that `publicPath` is a string prefix, but I can stop making a big deal about it if folks think that in practice, it's always joined with a path separator.

goldhand replied:
I'll just make note to make sure the documentation is clear on the path logic incase there is any trouble. We can revisit using `url.resolve()` if this is an issue.

anilanar commented on packages/workbox-webpack-plugin/src/lib/get-manifest-entries-with-webpack.js:
```diff
@@ -0,0 +1,137 @@
+    ), {});
+
+/**
+ * Generate an array of manifest entries using webpack's compilation data
```
Question: For dev environments, should workbox filter out source maps? Also, some people may be creating source maps even for production builds to debug deployed code. Some users may need to filter out some assets depending on their use cases.

goldhand replied:
See https://github.com/GoogleChrome/workbox/pull/808/files/a2dcaf4235f9f10456988a78759eb17dfedf6823#diff-dff2a0e4847723c3a1d299d74db40c0cR39

The idea was to provide a config option (`assetIgnoreRegex` or something) where users can provide an array of regular expressions to filter asset filenames out. By default we would use: [/\.map$/, /asset-manifest\.json$/]

It feels a little odd filtering by asset filenames. Normally (in my experience) webpack configs include/exclude assets based on chunk names, not actual filenames but for the case with `*.map$` files, they would need to be filtered by asset file names.

anilanar replied:
I guess source-maps can be filtered out by default. However I'm not sure how that would affect development experience when testing offline caching. Need to test it.

This problem is similar to how loaders handle file inclusion/exclusion using three properties: `{test, include, exclude}` and the code for that is in https://github.com/webpack/webpack/blob/master/lib/ModuleFilenameHelpers.js, look for `matchObject`. That's what `uglifyjs-webpack-plugin` is also using in  https://github.com/webpack-contrib/uglifyjs-webpack-plugin/blob/master/src/index.js.

goldhand replied:
So considering we filter both **chunkname** and **filename** what should the api look like?

We could provide explicit chunk / filename filtering eg:
```js
WorkboxPlugin({
  ...
  filterChunks: {
    test: /\.js($&#124;\?)/i,
  },
  filterFilenames: {
    exclude: /\/excludes/,
  },
});
```

Or we can just filter both using the same properties, eg:
```js
WorkboxPlugin({
  ...
  test: /\.js($&#124;\?)/i,
  exclude: /\/excludes/,
});
```
and just filter both **chunknames** and **filenames** by those provided values?

Something else to consider is [html-webpack-plugin's config](https://github.com/jantimon/html-webpack-plugin#configuration) the `{chunks, excludeChunks}` options are very nice. Maybe use those and apply the `{test, include, exclude}` options *only* to **filenames** (**chunknames** are handled by the `chunks` option...

goldhand replied:
@anilanar ^ (meant to tag you in that comment)

jeffposnick replied:
FYI: I've opened https://github.com/GoogleChrome/workbox/issues/935 to track this, and moved over some relevant conversation. I don't want too much of the conversation to be lost in a closed PR.

anilanar commented:
Great work! I'll retest after `HtmlWebpackPlugin` is handled.

anilanar commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,181 @@
+     * During the make phase of the webpack compilation, we use
+     * workbox-webpack-plugin/utils/copy-workbox-sw to add a built version of
+     * workbox-sw to the webpack compilation assets array.
+     */
```
Hmm, I think this is not what everyone would want. 
We build `service-worker` separately by another webpack build, using workbox from npm which is embed into service-worker file. We have bunch of custom loaders that modifies/optimizes/analyzes code, so we would not prefer to have it this way. This approach is also uncommon in webpack/js community.

goldhand replied:
Specifying the path to the service worker you create in `swSrc` would use that service worker then prepend it with `importScripts('workbox-sw.prod.v2.x.x.js', 'file-manifest.js')` so then you can just run:

`workboxBuild.precache(__precacheManifest)` in your service worker.

Not sure if this is the best flow...

see https://github.com/GoogleChrome/workbox/pull/808/files/a2dcaf4235f9f10456988a78759eb17dfedf6823#diff-5b99f4463e518de9c1c1cc612e8f6b59R73 for the swSrc parameter

anilanar replied:
I assumed a raw `service-worker.js` would look as follows:

```
import WorkboxSW from 'workbox-sw';
const workboxSW = new WorkboxSW();
workboxSW.precache(self.__precacheManifest);
```

After building with webpack and processing with workbox:

```
importScripts('file-manifest.{hash}.js');
(function () {
      /* webpack generated code here */
});
```

Ideally, users can opt to use `workbox-precaching` package instead of the whole `workbox-sw` when they don't need anything else than simple precaching.

jeffposnick replied:
FWIW, we're expecting it to be common for developers to refer to `workbox-sw` from our new CDN URL, support for which is baked in to `workbox-build` v3. The CDN `workbox-sw` basically acts as a lightweight loader + API surface, and will use `importScripts` at runtime to pull in additional dependencies for things like routing, caching, etc.

That being said, we know that not every developer will want to rely on the CDN copy, and Webpack users in particular are in a good position to bundle an optimized copy of `workbox-sw` + the additional routing, caching, etc. code and serve it locally.

CC: @philipwalton who's also exploring the ergonomics of bundling Workbox code.

goldhand replied:
This will need some more consideration. I think we need to decide if `workbox-sw.js` should be copied into the webpack assets array, use the CDN url or expect the developer to import the `workbox-sw` file into a `service-worker` webpack entry point. Maybe provide an option in the config to specify what option we should use and select a smart default... not sure what *smart* would be ;)

goldhand commented on packages/workbox-webpack-plugin/src/index.js:
```diff
@@ -0,0 +1,181 @@
+ *
+ * @module workbox-webpack-plugin
+ */
+class WorkboxWebpackPlugin {
```
Makes sense, originally had it that way but I didn't see it in any other workbox api so I removed it. @jeffposnick what do you think about having `manifestVarName` as an option that defaults to `__precacheManifest`?

Only problem is that in the `copy-service-worker` method, we would need to be able to update that param when `swBuild.precache(<manifestVarName)` is run

jeffposnick replied:
The format is an array of [`ManifestEntry` objects](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-build#.ManifestEntry).

One pattern that you could adopt for the standalone manifest imports is

```js
self.__precacheManifest = [entriesDefinedInThisFile].concat(self.__precacheManifest || []);
```

That's basically what we're already doing in the `v3` version of the SW template file, to merge anything passed in via the `manifestEntries` template variable with anything defined already via `self.__precacheManifest`.

goldhand commented on packages/workbox-webpack-plugin/src/lib/generate-or-copy-sw.js:
```diff
@@ -0,0 +1,47 @@
+    // use workbox-build to generate the service worker
+    return resolve(generateSWString(config));
+  } else {
+    return readFile(swSrc).then((serviceWorkerSource) =>
```
I prefer absolute paths, how would we enforce that?

anilanar replied:
`path.isAbsolute` looks useful. Plugin options would be validated in the constructor, similar to how webpack does it.

jeffposnick commented:
Thanks for your work on this, @goldhand, and thanks for your feedback, @anilanar!

I plan on giving it a thorough review and address some of the `workbox-build` comments. As discussed, given the amount of work in this PR, we're open to the idea of merging into `v3` as long as the core functionality is there, and track some incremental changes or outstanding feature requests in issues to be addressed in follow-up pRs.

jeffposnick commented:
I'm going to merge, and have filed separate issues to track some unresolved questions/feedback to be addressed in follow-up PRs:

- https://github.com/GoogleChrome/workbox/issues/933
- https://github.com/GoogleChrome/workbox/issues/934
- https://github.com/GoogleChrome/workbox/issues/935

There will also be some general code style and `workbox-build` integration PRs post-merge.

dingquantracy commented:
When will you guys publish the V3 to npm, so I can use the latest version to solve problems in webpack-dev-middle, thx
